zerver/tests/test_send_email.py

```diff
@@ -70,39 +74,132 @@ def test_limited_to_length(self) -> None:
+            backend, "_open", wraps=backend._open
+        ) as _open_call, mock.patch.object(backend, "close", wraps=backend.close) as close_call:
+            backend.connection = mock.MagicMock(spec=SMTP)
+            backend.connection.noop.return_value = [250]
+            initialize_connection(backend)
+
+            self.assertEqual(_open_call.call_count, 1)
+            self.assertEqual(backend.connection.noop.call_count, 1)
+
+            # Test the old connection case when it was closed by the server
+            backend.connection.noop.return_value = [404]
+            initialize_connection(backend)
+
+            # 2 more calls to open and 1 call to close
+            self.assertEqual(_open_call.call_count, 3)
+            self.assertEqual(close_call.call_count, 1)
+
+            # Test backoff procedure
+            _open_call.side_effect = OSError
+            with self.assertRaises(OSError):
+                initialize_connection(backend)
+            # 3 more calls to open as we try 3 times before giving up
+            self.assertEqual(_open_call.call_count, 6)
+
+    @time_machine.travel(timezone_now(), tick=False)
+    def test_max_connection_lifetime(self) -> None:
+        backend = PersistentSMTPEmailBackend()
+        backend.open()
+        self.assertEqual(backend.opened_at, timezone_now())
+
+        with mock.patch.object(
+            backend, "_open", wraps=backend._open
+        ) as _open_call, mock.patch.object(backend, "close", wraps=backend.close) as close_call:
+            # Old connection is open, but we still reopen a new connection because max connection lifetime is 0.
+            backend.connection = mock.MagicMock(spec=SMTP)
+            backend.connection.noop.return_value = [250]
             initialize_connection(backend)
-        # 3 more calls to open as we try 3 times before giving up
-        self.assertEqual(backend.open.call_count, 6)
+
+            self.assertEqual(close_call.call_count, 1)
+            self.assertEqual(_open_call.call_count, 2)
+            self.assertEqual(backend.opened_at, timezone_now())
+
+            # Old connection is open, we will not open a new connection because max connection lifetime is None.
+            with self.settings(EMAIL_MAX_CONNECTION_LIFETIME_IN_MINUTES=None):
+                # We have to set this a second time to the same values because
+                # backend.connection was overwritten in a previous
+                # `initialize_connection` call in this test.
+                backend.connection = mock.MagicMock(spec=SMTP)
+                backend.connection.noop.return_value = [250]
+                initialize_connection(backend)
+
+            self.assertEqual(close_call.call_count, 1)
+            # The increase in open.call_count by 1 is caused by us calling connection.open()
+            # to check whether an open connection already exists. In case of actually opening
+            # a new connection, the call count will increase by 2.
```

Comment:
It might be nicer to (thoughout, not just here) use `reset_mock()` so we can be explicit about the number of calls _here_, rather than having to diff to the previous part of the test.